### ISSUE TITLE
allow more than 1 PgBouncer replicas

### ISSUE THREAD

**Issue by low-on-mana** [airflow-helm/charts#622]
### Checks

- [X] I have checked for [existing issues](https://github.com/airflow-helm/charts/issues).
- [X] This report is about the [`User-Community Airflow Helm Chart`](https://github.com/airflow-helm/charts/tree/main/charts/airflow).

### Chart Version

latest

### Kubernetes Version

```shell
NA
```

### Helm Version

```shell
NA
```

### Description

We are using the latest version of this chart in production for airflow 2.3.0 ( we did this migration few days back ).

One of the issues we faced is related to pgbouncer. 
What happened was K8 rescheduled the pgbouncer pod to another node, since there is only 1 pod running we had one task failure which we had to retry manually later.

We can have safe_to_evict false or pod disruption budget as another solution but best would be to make pgbouncer HA by using multi pods. 

Can we have 2 pods for HA pgbouncer ?

```
spec:
  replicas: 1
  strategy:
    rollingUpdate:
      ## multiple pgbouncer pods can safely run concurrently
```
 https://github.com/airflow-helm/charts/blob/420eae29c454f6e7e6a7837706ca2e6c0fe792b8/charts/airflow/templates/pgbouncer/pgbouncer-deployment.yaml#L24

**Comment by thesuperzapper** [comment]
@low-on-mana @juroVee I agree that having multiple PgBouncer replicas would be (in theory) great for redundancy, especially during node outages/upgrades, the problem is that any disruption to the database connection during a transaction will result in airflow raising an error, which I doubt airflow will gracefully recover from. 

(NOTE: airflow uses [SQLAlchemy in "pessimistic" pooling mode](https://docs.sqlalchemy.org/en/14/core/pooling.html#dealing-with-disconnects) with the pre-ping approach, which can't handle mid-transaction failures)

That is to say, more PgBouncer replicas actually increases the possiblity of airflow trying to use a connection to a PgBouncer Pod that is no longer active (and crashing as a result).

We would need to investigate getting airflow to use a different SQLAlchemy pooling mode (to allow mid-transaction failures to be resolved gracefully) before we can increase PgBouncer replicas.

**Comment by kien-truong** [comment]
> (NOTE: airflow uses [SQLAlchemy in "pessimistic" pooling mode](https://docs.sqlalchemy.org/en/14/core/pooling.html#dealing-with-disconnects) with the pre-ping approach, which can't handle mid-transaction failures)

> We would need to investigate getting airflow to use a different SQLAlchemy pooling mode (to allow mid-transaction failures to be resolved gracefully) before we can increase PgBouncer replicas.

With PostgreSQL, it's not possible to recover from a broken connection mid-transaction, because a transaction is always tied to a specific db connection. In fact, I believe it's the same for MySQL as well, so there's really no way to improve the situation here.

> That is to say, more PgBouncer replicas actually increases the possiblity of airflow trying to use a connection to a PgBouncer Pod that is no longer active (and crashing as a result).

I don't think this is the case, SQLAlchemy do not switch between different PGBouncer during a connection lifetime: it just picks one of the PGBouncer to connect to whenever it need to add a new connection to the pool. 

At the time Airflow checks out a connection, if the `pgbouncer` pod is already dead, it will fail during pre-ping and simply be replaced with a different connection from a healthy PGBouncer: No crash will happen in this case.

However, increase the number of `pgbouncer` does increase the chance that one of them will fail after pre-ping, in the middle of a transaction. In that case, there's nothing Airflow can do but perform a retry, as is current.

**Comment by kien-truong** [comment]
On a side note, it's probably better to set `pgbouncer.safeToEvict` to `false` by default, since the loss of the pgbouncer pod can trigger cascade failures in a many other pods.
